### PR TITLE
Cache student data locally

### PR DESCRIPTION
**Pull request by AparaV**
## Why?
Right now, data is freshly fetched each time the user calls one of these methods: `info()`, `classes()`, `books()`, `gpa()`. This can be pretty slow if the internet connection is bad (like at public places). And this is bad because the same data is being fetched each time (the data we are retrieving does not change frequently).
So, I thought performance could be improved by caching these information locally inside the class the first time the method is called and retrieve them upon each subsequent call.

## Some numbers to support performance improvement
This was tested on a machine with:
- Python 3.5
- 40 mbps internet connection (download)

```bash
It took 309.355 ms to retrieve CUSession.info() without caching
It took 0.000 ms to retrieve CUSession.info() with caching

It took 684.469 ms to retrieve CUSession.classes() without caching
It took 0.000 ms to retrieve CUSession.classes() with caching
````
I couldn't time `CUSession.books()` and `CUSession.gpa()` because they didn't work. But I added a caching system to those as well.

## What if I want my program to fetch new data each time it's run?
No worries. You can now pass an optional argument `force=True` to any of the functions to forcefully reset your local cache. (By default `force` is set to `False` for all methods)

<hr>

The cache variables are currently private to the class. I can change this if necessary.

### CODE REVIEW

File: mycuinfo/cusession.py [hunk 1]
```diff
@@ -118,218 +118,256 @@ def __init__(self, username, password):
         # from here on, we are a regular user who has all the logged in cookies
         # we can do anything that a web user could (javascript not included)
         self.session = session
+        self.__student_info = None
+        self.__student_classes = {}
+        self.__student_books = {}
+        self.__student_gpa = None
```
Comment (mguida22) [inline mycuinfo/cusession.py]: It might be clearer to call these `self.__cached_info`, etc..

**Comment by mguida22** [comment]
I like the idea of this! You should probably add a note to the README about it so people are aware of it. Also the caching will only work so long as the program stays running, correct?

**Comment by biw** [comment]
Hi @AparaV,

Thanks for the pull request! Maybe I missed it @AparaV, but how does the cache know when to refresh itself other than manually?

File: mycuinfo/cusession.py (hunk 1 above)
Comment (AparaV) [reply mycuinfo/cusession.py]: @mguida22 yes, those variable names make more sense. Will refactor them now

**Comment by AparaV** [comment]
@mguida22 yes, you're right. The caching will work only as long as the program stays running. So, if I restart the program, I will have to fetch the info once again. And, I'll update the README to add a bit about caching

@719Ben right now, the cache does not refresh itself automatically (i.e., the user has to manually do that). I didn't think that was necessary because the information fetched is pretty much static. It does not change on a regular basis and will only get updated before semesters. And I guessed that not many want to keep their program alive for such long periods of time (instead they can use `cron jobs` to simulate that).

So I didn't want to incorporate automatic refresh of local data into this. But I can do that with the help of `time` module. I can start a timer each time the information is fetched freshly and then fetch it once again after a set interval of time (say 12 hours).

What do you think @719Ben ? Do we need to have the feature of automatic cache refreshing?

**Comment by mguida22** [comment]
I think it's best if the api doesn't refresh on it's own, and whoever implements this can deal with the details of when to refresh, since there's a lot of different ways you might want to handle when to reload data

**Comment by biw** [comment]
I agree with @mguida22, I guess my point is, if the library doesn't handle this for you, whats the point of including a cache in the codebase at all? Seems like an abstraction we should let users do.

**Comment by AparaV** [comment]
Yeah, I understand that @719Ben. So, I tried writing some code to allow the cache to refresh on it's own using `time`. But it soon turned out to be pretty messy and a bit convoluted.
Thus, I agree with you that we should just let the users handle that (it's a lot simpler for them) instead of the library natively providing a cache system.
Feel free to close the PR!

**Pull request closed without merging**